### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by bgurney**

### CODE REVIEW

File: src/Traverser.php [hunk 1]
```diff
@@ -76,13 +76,13 @@ public function children(): Collection
...
-    protected function inferChildren($objects): Collection
+    public function inferChildren($objects): Collection
     {
-        return collect($objects)
-            ->filter(function ($object) {
-                return $parent = static::make($object, $this->relations)->parent()
-                    && $this->is($parent);
-            });
+        return collect($objects)->filter(function ($object) {
+            $parent = static::make($object, $this->relations)->parent();
```
Comment (bgurney) [inline src/Traverser.php]: This is causing a recursive loop when $objects contains Objects with inferred parents.

Comment (bgurney) [reply src/Traverser.php]: Solved this by being more specific about what is passed into the inferParent and inferChildren methods.

- inferParent must only be passed objects that can actually be a parent
- inferChildren must only be passed objects that can actually be children

**Comment by sleicester** [comment]
It's hard to see the state of the setup of the tests, it's hidden in mock classes that also means you need to look into and understand the base Model class. Not that complicated, but perhaps we should just be using mocks here?

**Comment by bgurney** [comment]
I had wondered about using Mocks, but I'm not too sure whether that would be a better approach? I quite like how the test stubs mirror the documentation. How would you structure a Mock for `descendants()`?

**Pull request merged**